auto-scale promoted index

bhalevy opened the pull request:
Add column_index_auto_scale_threshold_in_kb to the configuration (defaults to 10MB).

When the promoted index (serialized) size gets to this
threshold, it's halved by merging each two adjacent blocks
into one and doubling the desired_block_size.

Fixes #4217

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

avikivity commented on sstables/mx/writer.cc:
```diff
@@ -571,11 +571,10 @@ class writer : public sstable_writer::writer_impl {
-        bytes_ostream offsets; // Serialized block offsets (uint32_t) relative to the start of "blocks".
-        uint64_t promoted_index_size = 0; // Number of pi_blocks inside blocks and first_entry;
+        // blocks is used for deferring serialization into blocks for small partitions.
+        utils::chunked_vector<pi_block> blocks;
```
This is a regression in memory consumption for small promoted indexes since each pi_block involves two allocations, and allocations are rounded  up to a power of two. bytes_ostream switches to chunks of size 128k pretty soon.

I suggest to drop this patch. It means that the size if not constant, but it should be enough to keep it in check.

bhalevy replied:
Hmm, my first thought was to reparse the existing serialized blocks in order to down-sample them and reserialize them into a new bytes_ostream, but this is complicated (and cpu consuming).

IIUC, what you suggest is variable resolution, so we keep using the same block ostream but double the desired block size so the additional blocks will use half the resolution.  We can then increase the current (per-partition write) threshold so the next auto-scaling point will be when the block ostream will grow in another threshold bytes in size.

avikivity replied:
Yes, access to the front of the partition will be faster than access into the back. But the motivation is to keep it simple and reduce chance of problems.

avikivity commented on sstables/mx/writer.cc:
```diff
@@ -840,6 +842,30 @@ void writer::add_pi_block() {
+            auto block2 = *it++;
+            block.last = std::move(block2.last);
+            block.width += block2.width;
+            block.open_marker = block2.open_marker;
```
I think you should keep block.marker here.

bhalevy replied:
I had the impression that it contains the most recent range tombstone when the block ends, not when it starts.
I.e. we need to keep the second one, not the first.  (Maybe this is an existing issue and it should be the last one before the block begins as its name suggests).

Refs:
1) https://github.com/scylladb/scylla/blob/6a3494442e6e870620b53adf8194fa845d4a5acc/sstables/mx/writer.cc#L828-L833

2) https://github.com/scylladb/scylla/blob/6a3494442e6e870620b53adf8194fa845d4a5acc/sstables/mx/writer.cc#L1342-L1346

If we want to keep the last `_current_tombstone` before the block begins we'll to keep a copy of it in the `_pi_block_m`.

avikivity replied:
No, the existing code and your change are correct.

avikivity commented on sstables/mx/writer.cc:
```diff
@@ -840,6 +842,30 @@ void writer::add_pi_block() {
+            block.width += block2.width;
+            block.open_marker = block2.open_marker;
+            write_pi_block(blocks, offsets, block, blocks.size());
+            new_blocks.emplace_back(std::move(block));
```
and yield, and deallocate the previous two blocks.

bhalevy replied:
yes (if we keep this patch)

bhalevy commented:
In v2:
- make desired_block_size per partition (reinitialized for every partition)
- add `column_index_auto_scale_threshold_in_kb` example to conf/scylla.yaml
  - and document what it does
- add `pi_auto_scale_events` metric.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/500/

avikivity commented on db/config.cc:
```diff
@@ -468,6 +468,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     /* Cache and index settings */
     , column_index_size_in_kb(this, "column_index_size_in_kb", value_status::Used, 64,
         "Granularity of the index of rows within a partition. For huge rows, decrease this setting to improve seek time. If you use key cache, be careful not to make this setting too large because key cache will be overwhelmed. If you're unsure of the size of the rows, it's best to use the default setting.")
+    , column_index_auto_scale_threshold_in_kb(this, "column_index_auto_scale_threshold_in_kb", value_status::Used, 10240,
```
Can you make this LiveUpdate? To avoid rolling restarts.

bhalevy replied:
Never done this, but I guess we can. Will do.

avikivity commented on sstables/sstables_manager.cc:
```diff
@@ -48,6 +48,7 @@ sstable_writer_config sstables_manager::configure_writer(sstring origin) const {
     sstable_writer_config cfg;
 
     cfg.promoted_index_block_size = _db_config.column_index_size_in_kb() * 1024;
+    cfg.promoted_index_auto_scale_threshold = _db_config.column_index_auto_scale_threshold_in_kb() * 1024;
```
This is sufficient for LiveUpdate - new writers will see updated values.

avikivity commented on sstables/mx/writer.cc:
```diff
@@ -844,6 +852,12 @@ void writer::add_pi_block() {
     ++_pi_write_m.promoted_index_size;
+
+    // auto-scale?
+    if (_pi_write_m.auto_scale_threshold && _pi_write_m.blocks.size() >= _pi_write_m.auto_scale_threshold) {
```
Better to use max for the threshold to avoid special values (0).

bhalevy replied:
Meaning tell the user to set to <size_t>::max() (or to an arbitrary high value) to effectively disable?

avikivity commented on test/boost/sstable_mutation_test.cc:
```diff
@@ -957,6 +957,44 @@ SEASTAR_TEST_CASE(test_promoted_index_blocks_are_monotonic) {
+        mt->apply(std::move(m));
+        sstable_writer_config cfg = env.manager().configure_writer();
+        cfg.promoted_index_block_size = 1;
+        cfg.promoted_index_auto_scale_threshold = 1024; // disable auto-scaling
```
1024 is not disabled

avikivity replied:
and we didn't test that auto-scaling works

bhalevy commented:
In v4 (33bad72fd2):
- Made column_index_auto_scale_threshold_in_kb `LiveUpdate`
- converted internally to `std::numeric_limits<size_t>::max()` when zero
  - removed test for special 0 value in `add_pi_block`
- fixed comment in test_promoted_index_blocks_are_monotonic_with_auto_scaling
  - and reduced the threshold even further to trigger auto-scaling earlier.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/503/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/503/

https://jenkins.scylladb.com/job/releng/job/Scylla-CI/503/artifact/testlog/x86_64_dev/mutation_test.test_query_digest.1123.log

Looks like this spurious failure: https://github.com/scylladb/scylla/issues/10554
Doesn't reproduce locally with same seed.

This PR does have 6fd1322bf3d9372d6adeb51af79f4957b81311d0, but as @denesb wrote there:
> This patch is a speculative fix.
So I'll reopen the issue.

Rebuilding CI...

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/533/
